File: src/leetcode/third/Code_041_FirstMissingPositive.java
package leetcode.third;

/**
 * 第一个缺失的正数
 * 时间复杂度O(n) 空间复杂度O(1)
 * <p>
 * 对于符合范围要求的nums[i], 映射到下标: nums[i] - 1, 例如 1 -> 0, 2 -> 1, 3 -> 2
 *
 * @since 2020-5-6 Wednesday 15:45
 */
public class Code_041_FirstMissingPositive {
    static int firstMissingPositive(int[] nums) {
        int len = nums.length;
        for (int i = 0; i < len; i++) {
            while (nums[i] > 0 && nums[i] <= len && nums[i] != nums[nums[i] - 1]) {
                swap(nums, i, nums[i] - 1);
            }
        }
        for (int i = 0; i < len; i++) {
            if (nums[i] != i + 1) {
                return (i + 1);
            }
        }
        return len + 1;
    }

    private static void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public static void main(String[] args) {
        System.out.println(firstMissingPositive(new int[]{3, 4, -1, 1}));
        System.out.println(firstMissingPositive(new int[]{1, 2, 0}));
        System.out.println(firstMissingPositive(new int[]{1, 1, 1, 1}));
        System.out.println(firstMissingPositive(new int[]{1, 2, 3}));
        System.out.println(firstMissingPositive(new int[]{0, 1, 2, 3}));
        System.out.println(firstMissingPositive(new int[]{0}));
        System.out.println(firstMissingPositive(new int[]{1}));
        System.out.println(firstMissingPositive(new int[]{10}));
        System.out.println(firstMissingPositive(new int[]{99, 100}));
        System.out.println(firstMissingPositive(new int[]{1, 100}));
        System.out.println(firstMissingPositive(new int[]{1, -1}));
    }
}

File: src/cpp-leetcode/first/Code_041_FirstMissingPositive.cpp
#include <iostream>
#include <vector>

using std::vector;

/**
 * find first missing posotive number in array
 *
 * @since 2020-10-23 Friday 16:46
 */
class Solution {
    void swap(vector<int>& nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

   public:
    int firstMissingPositive(vector<int>& nums) {
        int size = nums.size();
        for (int i = 0; i < size; i++) {
            while (nums[i] > 0 && nums[i] < size && nums[i] != nums[nums[i] - 1]) {
                swap(nums, i, nums[i] - 1);
            }
        }
        for (int i = 0; i < size; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        return size + 1;
    }
};

int main(int argc, char const* argv[]) {
    Solution s;
    vector<int> v1 = {7, 8, 9, 11, 12};
    vector<int> v2 = {3, 4, -1, 1};
    vector<int> v3 = {1, 2, 0};
    std::cout << s.firstMissingPositive(v1) << "\n";
    std::cout << s.firstMissingPositive(v2) << "\n";
    std::cout << s.firstMissingPositive(v3) << "\n";
}
